### ISSUE TITLE
Ability to export constant on app host

### ISSUE THREAD

**Issue by huoyaoyuan** [dotnet/core-setup#8901]
The use case is `NvOptimusEnablement`.
There seems to be a way to write export constants in IL, but IL cannot be written easily together with C#.
Using C++/CLI is unsuitable, because the whole project contains no native code and is cross platform.

A potential pure C# solution is adding a custom attribute:
```csharp
[AttributeUsage(AttributeTargets.Assembly)]
public sealed class ExportConstantAttribute : Attribute
{
    public string Name { get; set; }
    public Type Type { get; set; }
    public object Value { get; set; }
}

// usage
[assembly: ExportConstant("NvOptimusEnablement" ,typeof(int), 1)]
```
The build process emits the export on output native host.

Alternative:
```csharp
[AttributeUsage(AttributeTargets.Field)]
public sealed class DllExportAttribute : Attribute {}

[DllExport]
public const int NvOptimusEnablement = 1;
```
Let the compiler special treats it like `DllExport`, and just copy exports from managed to native.

**Comment by vitek-karas** [comment]
@jkoritzinsky 

I assume this is about exporting a native symbol with a specified value. And I assume the expectation is that this would show up on the `app.exe`. Is that correct?

**Comment by huoyaoyuan** [comment]
You're right. In .NET Core we are having less control of the native exe now.

**Comment by vitek-karas** [comment]
.NET Core in general doesn't support the `DllExport` functionality (be it with that attribute or any other). In general though, the ask so far was to support it on a "library" - so `.dll` on windows. And mainly about function exports, not constants.

Just to set the right expectations, even if we would enable `DllExport` like functionality, it is unlikely we would do so on the executable (`.exe`), mainly due to the way the executable is produced currently, which would make it rather hard to do this.

Is this something you need to support in some way on non-Windows?

The currently available solution to this would be your own native host. A relatively small (C++ probably) native project, which can load and run the app. Using the [native hosting](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/native-hosting.md) APIs it is pretty simple to write a native executable which loads the runtime and runs an app. With that you would have full control over the executable, including native exports. I understand that this would mean a need for a "native project" in your otherwise managed-only solution though.

**Comment by huoyaoyuan** [comment]
It's mostly about nvidia optimus. I don't know the nvidia driver behaviour on other platforms.
I've considered using C++/CLI as the host. It's should work, but is not fitting to the dependency and workflow of the repository.

Alternatively, nvidia says static importing `nvapi.dll` will also take the functionality. I tried with `DllImport`, but dumpbin says it's not imported on native host.

In conclusion: nvidia always require something on native host, but current build process can never emit such things. This is really painful.

We are making a game using OpenTK/SDL. Unity is not suitable because of the engine design.

**Comment by vitek-karas** [comment]
`DllImport` is always dynamic (will call `LoadLibrary` at runtime), so that's expected.

Currently the only customization the SDK offers for the app host is Win32 resources unfortunately.

I personally would not go with C++/CLI, for one I think we don't support C++/CLI executables (only libraries) - @jkoritzinsky should be able to confirm - and second it feels a bit too complex for the needs here.
